state/runtime/tracer/calltracer/call_tracer.go

```diff
@@ -0,0 +1,156 @@
+package calltracer
+
+import (
+	"math/big"
+	"sync"
+
+	"github.com/0xPolygon/polygon-edge/helper/hex"
+	"github.com/0xPolygon/polygon-edge/state/runtime/tracer"
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+var (
+	callTypes = map[int]string{
+		0: "CALL",
+		1: "CALLCODE",
+		2: "DELEGATECALL",
+		3: "STATICCALL",
+		4: "CREATE",
+		5: "CREATE2",
+	}
+)
+
+type Call struct {
+	Type    string  `json:"type"`
+	From    string  `json:"from"`
+	To      string  `json:"to"`
+	Value   string  `json:"value,omitempty"`
+	Gas     string  `json:"gas"`
+	GasUsed string  `json:"gasUsed"`
+	Input   string  `json:"input"`
+	Output  string  `json:"output"`
+	Calls   []*Call `json:"calls,omitempty"`
+
+	parent   *Call
+	startGas uint64
+}
+
+type CallTracer struct {
+	call               *Call
+	activeCall         *Call
+	activeGas          uint64
+	activeAvailableGas uint64
+
+	cancelLock *sync.Mutex
+	reason     error
+	stop       bool
+}
+
+func NewCallTracer() *CallTracer {
+	return &CallTracer{
+		cancelLock: &sync.Mutex{},
+	}
+}
+
+func (c *CallTracer) Cancel(err error) {
+	c.cancelLock.Lock()
+	defer c.cancelLock.Unlock()
+
+	c.reason = err
+	c.stop = true
+}
+
+func (c *CallTracer) cancelled() bool {
+	c.cancelLock.Lock()
+	defer c.cancelLock.Unlock()
+
+	return c.stop
+}
+
+func (c *CallTracer) Clear() {
+	c.call = nil
+	c.activeCall = nil
+}
+
+func (c *CallTracer) GetResult() (interface{}, error) {
+	return c.call, nil
+}
```

Comment:
Maybe we should propagate reason here, based GetResult implementation in the struct tracer implementation, WDYT?
Otherwise we should probably remove `reason` field altogether.

```suggestion
func (c *CallTracer) GetResult() (interface{}, error) {
    if c.reason != nil {
        return nil, c.reason
    }

	return c.call, nil
}
```

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/2001/commits/786e10e03573e7a20dd6ec34eb45ff613da5909c